xwiki-platform-distribution/xwiki-platform-distribution-migrations/src/main/java/org/xwiki/migrations/internal/R170400000XWIKI23160DataMigration.java

```diff
@@ -0,0 +1,175 @@
+        BaseObject xwikiPreferencesXObject =
+            xwikiPreferencesDocument.getXObject(xwikiPreferencesDocumentReferenceNoLocal);
+        BaseObject xwikiPreferencesXObjectFromXar =
+            xwikiPreferencesDocumentFromXar.getXObject(xwikiPreferencesDocumentReferenceNoLocal);
+        boolean metaChanged = clearField(xwikiPreferencesXObject, xwikiPreferencesXObjectFromXar, META_FIELD);
+
+        if (metaChanged) {
+            wiki.saveDocument(xwikiPreferencesDocument,
+                "[UPGRADE] empty field [%s] because it matches the default values".formatted(META_FIELD), xWikiContext);
+        }
+    }
+
+    private XWikiDocument loadFromXar(DocumentReference xwikiPreferencesDocumentReference) throws DataMigrationException
+    {
+        // We need to cast to be able to access to the XAR-specific 'getXarInstalledExtensions' method.
+        if (this.installedExtensionRepository
+            instanceof XarInstalledExtensionRepository xarInstalledExtensionRepository)
+        {
+            Collection<XarInstalledExtension> xarInstalledExtensions =
+                xarInstalledExtensionRepository.getXarInstalledExtensions(xwikiPreferencesDocumentReference);
+            XarInstalledExtension xarInstalledExtension = xarInstalledExtensions.stream()
+                .filter(
+                    it -> Objects.equals(it.getId().getId(), "org.xwiki.platform:xwiki-platform-distribution-ui-base"))
+                .findFirst()
+                .orElse(null);
+            if (xarInstalledExtension != null) {
+                ExtensionId extensionId = xarInstalledExtension.getId();
+                try {
+                    XarEntry localDocumentReference =
+                        new XarEntry(xwikiPreferencesDocumentReference.getLocalDocumentReference(),
+                            "XWiki/XWikiPreferences.xml");
+                    return this.packager.getXWikiDocument(xwikiPreferencesDocumentReference.getWikiReference(),
+                        localDocumentReference, xarInstalledExtension);
+                } catch (IOException | XarException e) {
+                    throw new DataMigrationException(
+                        "Failed to load [%s] from extension [%s]".formatted(xwikiPreferencesDocumentReference,
+                            extensionId), e);
+                }
+            } else {
+                // No matching extension found
+                return null;
+            }
+        } else {
+            throw new DataMigrationException(
+                "The resolved [%s] component cannot be cast to [%s]".formatted(InstalledExtensionRepository.class,
+                    XarInstalledExtensionRepository.class));
+        }
+    }
+
+    private boolean clearField(BaseObject object, BaseObject objectFromXar, String field)
+    {
+        String template = object.getStringValue(field);
+        if (StringUtils.isNotEmpty(template)) {
```

Comment:
I'm wondering if it would not make sense from performances point of view to check that earlier, before even loading the XAR version, for example.

Reply:
Same here, I'm not sure of the tradeoff between ease of read and perfs (which feels marginal, but I can be wrong). Checking for an option nonetheless.